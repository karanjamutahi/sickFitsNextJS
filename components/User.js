import React, { Component } from 'react';
import PropTypes from 'prop-types';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';


const CURRENT_USER_QUERY = gql`
    query CURRENT_USER_QUERY {
        me{
            id
            email
            firstname
            orders{
                id
            }
            cart {
                id
                quantity
                item {
                    id
                    price
                    image
                    title
                    description
                }
            }
        }
    }
`;

const User = (props) => (
    <Query query={CURRENT_USER_QUERY} {...props}>
        { 
            (payload) => {
                return props.children(payload);
            } 
        }
    </Query>
);

//Render Prop whose only child can be a function 
User.propTypes = {
    children: PropTypes.func.isRequired,
};
 
export default User;
export { CURRENT_USER_QUERY };