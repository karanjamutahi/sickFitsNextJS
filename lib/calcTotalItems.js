function totalItems(cart) {
    return cart.reduce((tally, cartItem) => tally + cartItem.quantity, 0)
}

export default totalItems;