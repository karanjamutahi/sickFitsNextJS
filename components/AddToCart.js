import React from 'react';
import gql from 'graphql-tag';
import { Mutation } from 'react-apollo';

import { CURRENT_USER_QUERY } from './User';

const ADD_TO_CART_MUTATION = gql`
    mutation addToCart($id: ID!){
        addToCart(id: $id) {
            id
            quantity
        }
    }
`;

class AddToCart extends React.Component {
    render(){
        const { id } = this.props;
        return(
            <Mutation mutation={ADD_TO_CART_MUTATION} refetchQueries={[{ query: CURRENT_USER_QUERY}]} variables={
                {
                    id,
                }}>
                {(addToCart, { loading, error }) => {
                    return (
                        <button onClick={addToCart}>Add To Cart ðŸ›’</button>
                    )
                }}
            </Mutation>
            
        )
    }
}

export default AddToCart;