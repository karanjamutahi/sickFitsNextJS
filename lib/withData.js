import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { devEndpoint, prodEndpoint } from '../config';
import { LOCAL_STATE_QUERY } from '../components/Cart';

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? devEndpoint : devEndpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_, variables, { cache }) {
            //read cartopen value from cache
             const { cartOpen } = cache.readQuery({
               query: LOCAL_STATE_QUERY,
             });
             //Write the cart State to the opposite
             const data = {
               data: {
                 cartOpen: ! cartOpen
               }
             };
             cache.writeData(data);
             return data; 
          }
        }
      },
      defaults: {
        cartOpen: false,
      }
    }
  });
}

export default withApollo(createClient);
