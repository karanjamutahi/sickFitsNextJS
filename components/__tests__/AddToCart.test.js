import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import Cart, { LOCAL_STATE_QUERY } from '../Cart';
import { CURRENT_USER_QUERY } from '../User';
import { fakeUser, fakeCartItem } from '../../lib/testUtils';
import { ApolloConsumer } from 'react-apollo';
import AddToCart, {ADD_TO_CART_MUTATION} from '../AddToCart';
import { ApolloClient } from 'apollo-boost';            

const mocks = [
    {
        request: { query: CURRENT_USER_QUERY },
        result: { data: { me: {
            ...fakeUser(),
            cart: [],
        }}}
    },
    {
        request: { query: CURRENT_USER_QUERY },
        result: { data: { me: {
            ...fakeUser(),
            cart: [fakeCartItem()],
        }}}
    },
    {
        request: { query: ADD_TO_CART_MUTATION, variables: { id: 'abc123' }},
        result: { data: {
            addToCart: {
                ...fakeCartItem(),
                quantity: 1,
            }
        }}
    },
    
]

describe('AddToCart Component', () => {
    it("Renders and Matches Snapshot", async () => {
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <AddToCart id="abc123"></AddToCart>
            </MockedProvider>
        );
        await wait();
        wrapper.update();
        //console.log(wrapper.debug());
        expect(toJSON(wrapper.find('button'))).toMatchSnapshot();
    });

    xit("adds an item when clicked", async () => {
        let apolloClient;
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
               <ApolloConsumer>
                   {(client) => {
                       apolloClient = client;
                       return(
                           <AddToCart id="abc123"></AddToCart>
                       )
                   }}
               </ApolloConsumer>
            </MockedProvider>
        );
        await wait();
        wrapper.update();
        const  { data: { me }}  = await apolloClient.query({ query: CURRENT_USER_QUERY });
        //console.log(me);
        expect(me.cart).toHaveLength(0);
        const button = wrapper.find('button');
        //console.log(button.debug());
        button.simulate('click');
        await wait();
        const  { data: { me: me2 }}  = await apolloClient.query({ query: CURRENT_USER_QUERY });
        //console.log(me2);
        expect(me2.cart).toHaveLength(1);
        expect(me2.cart[0].id).toBe('omg123');
        expect(me2.cart[0].quantity).toBe(3);
    });

    it("Changes from Add to Adding when clicked", async () => {
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <AddToCart id="abc123"></AddToCart>
            </MockedProvider>
        );

        await wait();
        wrapper.update();
        expect(wrapper.text()).toContain('Add To Cart');
        wrapper.find('button').simulate('click');
        expect(wrapper.text()).toContain('Adding To Cart');
    });
});