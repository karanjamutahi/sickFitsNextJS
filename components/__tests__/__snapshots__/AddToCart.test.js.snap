// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddToCart Component Renders and Matches Snapshot 1`] = `
<button
  onClick={[Function]}
>
  Add
   To Cart ðŸ›’
</button>
`;
